diesel_derives/src/sql_function.rs

```diff
@@ -1,25 +1,172 @@
+
+    let sql_name = parse_sql_name_attr(&mut input).unwrap_or_else(|| input.fn_name.to_string());
+
+    input.fn_name = format_ident!("{}_{}", input.fn_name, variant_no);
+
+    let nonvariadic_args_count = input.args.len() - variadic_argument_count;
+
+    let variadic_type_params: Vec<_> = input
+        .generics
+        .type_params()
+        .skip(nonvariadic_args_count)
+        .cloned()
+        .collect();
+
+    let variadic_args: Vec<_> = input
+        .args
+        .iter()
+        .skip(nonvariadic_args_count)
+        .cloned()
+        .collect();
+
+    input
+        .generics
+        .type_params_mut()
+        .skip(nonvariadic_args_count)
+        .for_each(|type_param| {
+            type_param.ident = format_ident!("{}1", type_param.ident);
+        });
+
+    for additional_generic in 2..=variant_no {
+        for mut type_param in variadic_type_params.clone() {
+            type_param.ident = format_ident!("{}{}", type_param.ident, additional_generic);
+            input.generics.params.push(GenericParam::Type(type_param));
+        }
+    }
+
+    input
+        .args
+        .iter_mut()
+        .skip(nonvariadic_args_count)
+        .for_each(|arg| {
+            append_index_to_strict_fn_arg(arg, 1);
+        });
+
+    for additional_arg in 2..=variant_no {
+        for mut arg in variadic_args.clone() {
+            append_index_to_strict_fn_arg(&mut arg, additional_arg);
+            input.args.push(arg);
+        }
+    }
+
+    expand_nonvariadic(input, sql_name, legacy_helper_type_and_module)
+}
+
+fn add_variadic_doc_comments(attributes: &mut Vec<Attribute>, fn_name: &str) {
+    let mut doc_comments_end = attributes.len()
+        - attributes
+            .iter()
+            .rev()
+            .position(|attr| match &attr.meta {
+                Meta::NameValue(MetaNameValue { path, .. }) => path.is_ident("doc"),
+                _ => false,
+            })
+            .unwrap_or(attributes.len());
+
+    let fn_family = format!("`{0}_1`, `{0}_2`, ... `{0}_n`", fn_name);
+
+    let doc_comments: Vec<Attribute> = parse_quote! {
+        ///
+        /// # Variadic functions
+        ///
+        /// This function is variadic in SQL, so there's a family of functions
+        /// on a diesel side:
+        ///
+        #[doc = #fn_family]
+        ///
+        /// Here, the postfix indicates repetitions of variadic arguments.
+        /// To use this function, the appropriate version with the correct
+        /// argument count must be selected.
+    };
+
+    for new_attribute in doc_comments {
+        attributes.insert(doc_comments_end, new_attribute);
+        doc_comments_end += 1;
+    }
+}
+
+fn append_index_to_strict_fn_arg(arg: &mut StrictFnArg, index: usize) {
+    arg.name = format_ident!("{}_{}", arg.name, index);
+
+    let Type::Path(mut ty_path) = arg.ty.clone() else {
+        panic!("Variadic arguments are expected to have path type");
```

Comment:
I would prefer returning a `syn::Error` with the correct span here to get better error messages